src/nodes/hypertable_modify.c

```diff
@@ -93,6 +110,224 @@ get_chunk_dispatch_states(PlanState *substate)
 	return NIL;
 }
 
+#if PG14_GE
+/*
+ * Helper method which returns true if SEGMENT BY column
+ * is NULLABLE. This method also returns the attribute
+ * number of the SEGMENT BY column in compressed chunk.
+ */
+static bool
+is_segmentby_col_null(Chunk *curr_comp_chunk, char *column_name, AttrNumber *attrno)
+{
+	bool is_null = false;
+	HeapTuple tp;
+	/*
+	 * check if SEGMENT BY column is NULLABLE or not
+	 */
+	tp = SearchSysCacheAttName(curr_comp_chunk->table_id, column_name);
+	if (HeapTupleIsValid(tp))
+	{
+		Form_pg_attribute att_tup = (Form_pg_attribute) GETSTRUCT(tp);
+		*attrno = att_tup->attnum;
+		is_null = att_tup->attnotnull;
+		ReleaseSysCache(tp);
+	}
+	else
+		elog(ERROR,
+			 "Could not find column: \"%s\" in table: \"%s\"",
+			 column_name,
+			 get_rel_name(curr_comp_chunk->table_id));
+	return !is_null;
+}
+
+/*
+ * Helper method which returns true if given column_name
+ * is configured as SEGMENT BY column in a compressed hypertable
+ */
+static bool
+is_segmentby_col(List *ht_compression_info, char *column_name)
+{
+	ListCell *lc;
+	foreach (lc, ht_compression_info)
+	{
+		FormData_hypertable_compression *fd = lfirst(lc);
+		if (namestrcmp(&fd->attname, column_name) == 0)
+		{
+			if (fd->segmentby_column_index > 0)
+				return true;
+			break;
+		}
+	}
+	return false;
+}
+
+/*
+ * This method will evaluate the expression (node), extract
+ * left and right operands, check if one of the operands is
+ * a simple Var type. If its Var type extract its corresponding
+ * column name from hypertable_compression catalogue table.
+ * If extracted column is a SEGMENT BY column then build
+ * scan keys.
+ */
+static bool
+check_predicate(Node *node, Chunk *curr_chunk, Chunk *curr_comp_chunk, ScanKeyData **scankey,
+				Bitmapset **null_columns, int *is_null_checks)
+{
```

Comment:
Is it possible to do this once per hypertable? This looks costly, because we're redoing mostly the same work for every chunk, and there can easily be thousands of chunks. Can we compute the nullable columns and scankeys once for hypertable, and then for each chunk just translate the attnos?

Reply:
Sounds good. Now i only evaluate the predicates to check if there are any SEGMENT BY columns and what type of condition it is.
ex: <segmentby_col> = <Const> OR  <segmentby_col> IS NULL OR  <segmentby_col> IS NOT NULL

In later code where i iterate over the chunks, i build scan keys and decompress compressed chunk.

Reply:
But the current code still traverses the RestrictInfos once per chunk, right? We could do this once per hypertable, but I think we can leave this optimization for the future.

Reply:
Nope, please check the function call to 
```
static CompressedBatchPredicates *fill_predictate_context(Chunk *ch, List *predicates)
```
I pass 1 of the chunks to evaluate WHERE and save all context in CompressedBatchPredicates.
This saved context is used during processing of each chunk where i fill the attribute numbers for building scan keys.

So the optimisation you proposed is still in place.